metadata-ingestion/src/datahub/utilities/metrics.py

```diff
@@ -0,0 +1,80 @@
+from abc import abstractmethod
+from enum import Enum
+from logging import Logger, getLogger
+from typing import Any, List, Optional
+
+logger: Logger = getLogger(__name__)
+
+
+class DatahubCustomEnvironment(Enum):
```

Comment:
This is specific to Affirm, so if someone else wanted to use the CloudWatchMetricReporter. Then this won't make sense for them.

I feel like the metrics stuff is something the community may be interested in and we can contribute it back.

Reply:
This would be tricky to contribute back in a clean way I think. For metrics I like keeping things typed (i.e don't accept arbitrary strings) to prevent accidental metric explosion (which can impact AWS service limits and cause potentially high costs).

Ideally Python could support some type of enum inheritance like this:
```
class DatahubCustomMetric(Enum):
# empty

class #DatahubCustomEnvironmen(Enum):
# empty

# another file (Affirm specific, not contributed back)
class AffirmDatahubCustomMetric(DatahubCustomMetric):
# all the definitions we have already

class AffirmDatahubCustomEnvironment(DatahubCustomEnvironment):
# all the definitions we have already
```

This way we could have the metrics lib still enforce that values are from enums. But it isn't possible as far as I know.

So, if we want to make this generic enough to contribute back, I think we'd have to make the library functions take in strings. Also, the use-cases for needing something like this are limited just to cron jobs, so I'm not sure if it'd add much value to the community (AFAIK all other Datahub components are long running services which use prometheus)

Reply:
Agreed strongly typed better than strings.

> Also, the use-cases for needing something like this are limited just to cron jobs, so I'm not sure if it'd add much value to the community (AFAIK all other Datahub components are long running services which use prometheus).

Yeah, the use case might be limited to just us, but I was thinking in general, if we can make changes contributable back without much more effort, it'd be the better way to go.
And I guess the `metadata-ingestion` portion of the repo are not long running services, it's for ingestion which can be run as jobs.


But zooming out on just the metrics gathering:
How much value do we get for identifying Live vs Analytics? Ideally, we know the dataplatform i.e. mysql, snowflake, etc? rather than a specific environment?

Reply:
Without including an environment dimension for live vs analytics, metrics would be combined for the two environments (i.e CW graphs and alarms couldn't distinguish between the two). I think having them separated makes it easier to see/isolate trends and weird behaviors

Reply:
So I mean comparing include the data platform which is more specific instead of environment? Snowflake, MySQL, etc. And data platform's cardinality shouldn't be too big.

Reply:
Ah I see. Yeah that's a good idea.